### DISCUSSION TITLE
OQS_Provider with MQTT

### DISCUSSION THREAD

**Discussion by NarjasHadjer** [open-quantum-safe/oqs-provider#433] [Q&A]
I am working on integrating OQS_Provider with MQTT on a Raspberry Pi 3B+. I have created the CA and the server certificate, and I have signed it with the CA. However, when I start Mosquitto, it returns the error: OpenSSL Error[0]: error:0A0000F7
routines::unknown certificate type (Sorry, I can't take screenshots on the Raspberry Pi because it's buggy, but I will provide the openssl.cnf and mosquitto.cnf configuration files and i took pictures with my phone
![Messenger_creation_56c1a636-79c9-46b9-95d8-d159a31102a5](https://github.com/open-quantum-safe/oqs-provider/assets/172641263/de2a2d8a-fa9d-4b83-93ce-092f68fc8d19)
![Messenger_creation_834192a4-4c76-4885-a4f5-f49ae09895ec](https://github.com/open-quantum-safe/oqs-provider/assets/172641263/97fbf271-5c0c-4166-b5a0-9fa77a9cf77f)

openssl.cnf : i added these lines to the default configuration

[provider_sect]
legacy = legacy_sect
oqsprovider = oqsprovider_sect
[default_sect] 
activate = 1
[legacy_sect]
activate = 1
[oqsprovider_sect]
activate = 1
[openssl_init]
providers = provider_sect
ssl_conf = ssl_sect
[ssl_sect]
system_default = system_default_sect
[system_default_sect]
CipherString = DEFAULT:@SECLEVEL=2 ( i put here 0, cause i had an error : ca md too weak when i tried to run mosquitto first time)
Groups = kyber768:kyber1024

mosquitto.conf : I added these lines to the default configuration

istener 1883
listener 8883
protocol mqtt
cafile /home/narjas/certif/Certificate_CA.crt 
keyfile /home/narjas/certif/Server.key 
certfile /home/narjas/certif/Certificate.crt
use_identity_as_username true
require_certificate true

tls_version tlsv1.3

And even when i run the test given, i get the same error :
![a88943c9-67a9-420a-8090-185913a8baccphoto](https://github.com/open-quantum-safe/oqs-provider/assets/172641263/c4c33b1f-c495-414d-8eca-5c6d98741cf0)

**Comment by baentsch** [comment]
This error message indicates most likely that `oqsprovider` is not active. Please proceed as per [the bug reporting template](https://github.com/open-quantum-safe/oqs-provider/issues/new?assignees=&labels=bug&projects=&template=bug_report.md&title=) by asking your setup to report actual provider availability (`openssl list -providers`) and [ensure](https://github.com/open-quantum-safe/oqs-provider/blob/main/USAGE.md#activation) `oqsprovider` is listed. Maybe @aniket-rulez can provide hands-on help as per https://github.com/open-quantum-safe/oqs-provider/issues/443#issuecomment-2209194125.

**Reply by NarjasHadjer** [reply]
The OQS_Provider is active, i can use post-quantum algorithms to create keys, and sign data, certificates ...., but when i activate TLS on MQTT, the server fails to start, indicating the error : unknown certificates, i tried everything, but still get the same error !

**Reply by baentsch** [reply]
Have you looked at the Mosquitto demo integration code on oqs-demos to check for non-obvious config differences between your setup and the one there?

**Reply by NarjasHadjer** [reply]
Yes, but I think I’ll redo everything with a newer version of OpenSSL, as I’ve seen recommended in other discussions. In fact, there are even some algorithms I haven’t been able to use. Thank you for your response. If I encounter the same error again, it must be due to another cause.

**Reply by baentsch** [reply]
> I’ll redo everything with a newer version of OpenSSL

If you want to use TLS sigs, indeed 3.2 is the minimum.
